src/state.rs

```diff
@@ -122,40 +126,103 @@ impl StateContainer {
         }
     }
 
-    /// Fetch state from existing memory mapped file or create new one
+    /// Fetch state from existing memory mapped file or create new one.
     #[inline]
-    pub(crate) fn state(&mut self, create: bool) -> Result<&mut State, SynchronizerError> {
```

Comment:
Perhaps, instead of creating `state_read` and `state_write`, add new parameter `write_lock: bool`? Then most logic should remain the same / as is.

Reply:
I suspect this will incur the performance cost of an additional branch check. I've put a lot of effort into avoiding unnecessary runtime cost (including  ae17e94, which was a substantial rewrite to avoid a single branch for a hypothetical use-case where write locking was only enabled for some `Synchronizer`s).

If runtime cost is of little concern, perhaps you would prefer the implementation in 171227e?

Reply:
I don't mind `state_read` / `state_write` too much and would like best performance we can get out of this.
Last idea on this - maybe we can templatize `state` method with `const bool`, so that read/write implementations compiled without the branch?

Reply:
Implemented in e1bbaab. I'm slightly concerned about the ergonomics, but it works as intended:

Note, that due to guard lifetimes, this is not possible:
```rust
    #[inline]
    pub(crate) fn state<const WRITE: bool>(
        &'a mut self,
        create: bool
    ) -> Result<&'a mut State, SynchronizerError> {
        // ...
        let mmap = if WRITE {
            let mut guard = self.mmap.as_mut().unwrap().lock()?;
            guard.deref_mut()
           // guard is dropped here
        } else {
            self.mmap.as_ref().unwrap().read()
        }

        Ok(unsafe { &mut *(mmap.as_ptr() as *mut State) })
    }
```

Reply:
Commit https://github.com/cloudflare/mmap-sync/commit/e1bbaaba67981e98faf6dd865b98445c52e1a4ae looks good, ergonomics are acceptable and reduces code duplication.